pkg/controllers/managementuser/rbac/roletemplates/roletemplate_handler.go

```diff
@@ -0,0 +1,238 @@
+package roletemplates
+
+import (
+	"errors"
+	"reflect"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	crbacv1 "github.com/rancher/wrangler/v3/pkg/generated/controllers/rbac/v1"
+	"github.com/rancher/wrangler/v3/pkg/slice"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var promotedRulesForProjects = map[string]string{
+	"navlinks":          "ui.cattle.io",
+	"nodes":             "",
+	"persistentvolumes": "",
+	"storageclasses":    "storage.k8s.io",
+	"apiservices":       "apiregistration.k8s.io",
+	"clusterrepos":      "catalog.cattle.io",
+	"clusters":          "management.cattle.io",
+}
+
+const (
+	clusterRoleOwnerAnnotation = "authz.cluster.cattle.io/clusterrole-owner"
+	aggregationLabel           = "management.cattle.io/aggregates"
+	projectContext             = "project"
+	promotedSuffix             = "promoted"
+)
+
+func newRoleTemplateHandler(uc *config.UserContext) *roleTemplateHandler {
+	return &roleTemplateHandler{
+		crClient: uc.Management.Wrangler.RBAC.ClusterRole(),
+	}
+}
+
+type roleTemplateHandler struct {
+	crClient crbacv1.ClusterRoleController
+}
+
+// OnChange ensures that the following Cluster Roles exist:
+//  1. a ClusterRole with the same name as the RoleTemplate
+//  2. an Aggregating ClusterRole that aggregates all inherited RoleTemplates with the name "RoleTemplateName-aggregator"
+//
+// For RoleTemplates with the Context == "Project", we also ensure:
+//  1. If the RoleTemplate has any rules for Global Resources, make a ClusterRole with those named "RoleTemplateName-promoted"
+//  2. an Aggregating ClusterRole that aggregates all inherited RoleTemplates' promoted Cluster Roles named "RoleTemplateName-promoted-aggregator"
+func (rtl *roleTemplateHandler) OnChange(key string, rt *v3.RoleTemplate) (*v3.RoleTemplate, error) {
```

Comment:
I believe `OnChange` is also triggered on object removal. Since you are already using `OnRemove` for the cleanup mechanism, you need to explicitly early abort here if `rt == nil` or `rt.DeletionTimestamp != nil`